src/core/xmlHttpRequest.js

```diff
@@ -0,0 +1,113 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020-2024 (original work) Open Assessment Technologies SA ;
+ */
+import httpRequestFlowFactory from 'core/request/flowFactory';
+
+const XHR_READY_STATE_OPENED = 1;
+const XHR_READY_STATE_HEADERS_RECEIVED = 2;
+const XHR_READY_STATE_HEADERS_DONE = 4;
+
+/**
+ * XHR implementation of Fetch API
+ * @param {string} url
+ * @param {Object} options - fetch request options that implements RequestInit (https://fetch.spec.whatwg.org/#requestinit)
+ * @param {Function} [options.onUploadProgress]
+ * @param {Function} [options.onDownloadProgress]
+ * @returns {Promise<Response>}
+ */
+function xhr(url, options) {
+    return new Promise(resolve => {
+        const request = new XMLHttpRequest();
+        let responseBody = null;
+        const responseHeaders = new Headers();
+
+        if (typeof options.onUploadProgress === 'function') {
+            request.upload.addEventListener('progress', options.onUploadProgress);
+        }
+        if (typeof options.onDownloadProgress === 'function') {
+            request.addEventListener('progress', options.onDownloadProgress);
+        }
+        request.addEventListener('readystatechange', () => {
+            switch (request.readyState) {
+                case XHR_READY_STATE_OPENED:
+                    // eslint-disable-next-line no-case-declarations
+                    for (const header in options.headers) {
+                        request.setRequestHeader(header, options.headers[header]);
+                    }
+                    break;
+                case XHR_READY_STATE_HEADERS_RECEIVED:
+                    request.getAllResponseHeaders().trim().split(/[\r\n]+/).forEach((line) => {
+                        const parts = line.split(": ");
+                        const header = parts.shift();
+                        const value = parts.join(": ");
+                        if (header) {
+                            responseHeaders.append(header, value);
+                        }
+                    });
+                    break;
+                case XHR_READY_STATE_HEADERS_DONE:
+                    responseBody = request.response;
+                    // Response with null body status cannot have body
+                    if ([101, 204, 205, 304].includes(request.status)) {
+                        responseBody = null;
+                    }
+                    if (request.responseType === 'json') {
+                        responseBody = JSON.stringify(request.response);
+                    }
+                    // eslint-disable-next-line no-case-declarations
+                    const response = new Response(responseBody, {
+                        status: request.status,
+                        statusText: request.statusText,
+                        headers: responseHeaders
+                    });
+                    resolve(response);
+                    break;
```

Comment:
`break` is not needed. But do we need default just in case?

Reply:
No, `default` will result in other statuses going there (and it'll be empty anyway), but I'll keep the `break;` for consistency